future-apps/payroll/test/payroll_allowed_tokens.js

```diff
@@ -0,0 +1,95 @@
+const { assertRevert } = require('@aragon/test-helpers/assertThrow')
+
+const getContract = name => artifacts.require(name)
+const getEvent = (receipt, event, arg) => { return receipt.logs.filter(l => l.event == event)[0].args[arg] }
+
+contract('Payroll, allowed tokens,', function(accounts) {
+  const [owner, employee] = accounts
+  const {
+    deployErc20TokenAndDeposit,
+    addAllowedTokens,
+    redistributeEth,
+    getDaoFinanceVault,
+    initializePayroll
+  } = require("./helpers.js")(owner)
+
+  let payroll
+  let payrollBase
+  let priceFeed
+  let dao
+  let finance
+  let vault
+
+  let erc20Tokens = []
+  const ERC20_TOKEN_DECIMALS = 18
+  const SALARY = (new web3.BigNumber(1)).times(10 ** Math.max(ERC20_TOKEN_DECIMALS - 4, 1))
+  const RATE_EXPIRY_TIME = 1000
+  const NOW_MOCK = new Date().getTime()
+  let ETH
+  let MAX_ALLOWED_TOKENS
+
+  const MAX_GAS_USED = 7e6
+
+  before(async () => {
+    payrollBase = await getContract("PayrollMock").new()
+
+    const daoAndFinance = await getDaoFinanceVault()
+
+    dao = daoAndFinance.dao
+    finance = daoAndFinance.finance
+    vault = daoAndFinance.vault
+
+    priceFeed = await getContract("PriceFeedMock").new()
+    priceFeed.mockSetTimestamp(NOW_MOCK)
+
+    MAX_ALLOWED_TOKENS = await payrollBase.getMaxAllowedTokens()
+    // Deploy ERC 20 Tokens (0 is for ETH)
+    for (let i = 1; i < MAX_ALLOWED_TOKENS; i++) {
+      erc20Tokens.push(await deployErc20TokenAndDeposit(owner, finance, vault, `Token ${i}`, ERC20_TOKEN_DECIMALS))
+    }
+
+    // make sure owner and Payroll have enough funds
+    await redistributeEth(accounts, finance)
+
+    const etherTokenConstantMock = await getContract("EtherTokenConstantMock").new()
+    ETH = await etherTokenConstantMock.getETHConstant()
+  })
+
+  beforeEach(async () => {
+    payroll = await initializePayroll(
+      dao,
+      payrollBase,
+      finance,
+      erc20Tokens[0],
+      priceFeed,
+      RATE_EXPIRY_TIME
+    )
+
+    await payroll.mockSetTimestamp(NOW_MOCK)
+
+    // adds allowed tokens
+    await addAllowedTokens(payroll, erc20Tokens)
+
+    const startDate = parseInt(await payroll.getTimestampPublic.call(), 10) - 2628005 // now minus 1/12 year
+    // add employee
+    const receipt = await payroll.addEmployee(employee, SALARY, "Kakaroto", 'Saiyajin', startDate)
+  })
+
+  it('fails adding one more token', async () => {
+    const erc20Token = await deployErc20TokenAndDeposit(owner, finance, vault, 'Extra token', ERC20_TOKEN_DECIMALS)
+
+    return assertRevert(async () => {
```

Comment:
No need to be async, it can just be: `return assertRevert(() => payroll.addAllowedToken(erc20Token.address))` with proper indentation

Reply:
I've been thinking about this recently and my recommendation is that it's probably best as

```
await assertRevert(async () => {
  await ...
})
```

Since the chances of screwing up later (if we add more statements to the test) are lower if we're explicit about the `async` and `await`s.

Reply:
That makes sense